README.rst
===========
mqtt-codec
===========

A weapons grade MQTT packet encoder and decoder (codec).


Status
=======

The `mqtt-codec` package is an
`MQTT 3.1.1 <http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.pdf>`_
packet encoder and decoder (codec).  The library has high test coverage
(~94%) and is known to perform well in distributed IoT networks with
thousands of nodes.


Installation
=============

The mqtt-codec package can be from `<pypi.org>`_ with
`pip <https://pypi.org/project/pip/>`_:

.. code-block:: bash

   pip install mqtt-codec

Usage
======

An encode/decode cycle looks like this:

.. code-block:: python

   >>> from io import BytesIO
   >>> from binascii import b2a_hex
   >>> import mqtt_codec.packet
   >>> import mqtt_codec.io
   >>>
   >>> # Encode a Connect packet
   >>> will = mqtt_codec.packet.MqttWill(qos=0, topic='hello', message='message', retain=True)
   >>> connect = mqtt_codec.packet.MqttConnect(client_id='client_id', clean_session=False, keep_alive=0, will=will)
   >>> with BytesIO() as f:
   ...   num_bytes_written = connect.encode(f)
   ...   buf = f.getvalue()
   ...
   >>> assert len(buf) == num_bytes_written
   >>> print('0x{} ({} bytes)'.format(b2a_hex(buf), len(buf)))
   0x102500044d515454042400000009636c69656e745f6964000568656c6c6f00076d657373616765 (39 bytes)
   >>>
   >>> # Decode the connect packet and assert equality.
   >>> with mqtt_codec.io.BytesReader(buf) as f:
   ...   num_bytes_read, decoded_connect = connect.decode(f)
   ...
   >>> assert len(buf) == num_bytes_written
   >>> assert connect == decoded_connect
   >>> print('  Encoded {}'.format(connect))
     Encoded MqttConnect(client_id='client_id', clean_session=False, keep_alive=0, username=***, password=***, will=MqttWill(topic=hello, payload=0x6d657373616765, retain=True, qos=0))
   >>> print('= Decoded {}'.format(decoded_connect))
   = Decoded MqttConnect(client_id=u'client_id', clean_session=False, keep_alive=0, username=***, password=***, will=MqttWill(topic=hello, payload=0x6d657373616765, retain=True, qos=0))


Python Requirements
====================

The ``mqtt-codec`` project has been tested on Linux against these
environments:

* Python 2.7
* Python 3.4
* Python 3.5
* Python 3.6
* Python 3.7

Python versions Python 3.0 - 3.3 may work but are not tested as part of
the project continuous integration infrastructure.


Library Requirements
=====================

When running Python versions less than 3.4 the ``enum34`` package is
required.  There are no other package requirements.


Project Infrastructure
=======================

The project is coordinated through public infrastructure available at
several places:

* `Releases (pypi) <https://pypi.org/project/mqtt-codec>`_
* `Documentation (readthedocs.io) <https://mqtt-codec.readthedocs.io/en/latest/>`_
* `Bug Tracker (github) <https://github.com/kcallin/mqtt-codec/issues>`_
* `Code Repository (github) <https://github.com/kcallin/mqtt-codec>`_


CHANGELOG.rst
===========
Change Log
===========


1.0.2 (2019-01-27)
===================

Fix
----
#5: QoS=0 publish messages incorrectly read/write packet_id.

    In Publish messages with QoS=0, packet_id is being serialized and
    deserialized in violation of MQTT 3.1.1 spec (See 3.3.2.2).

    https://github.com/kcallin/haka-mqtt/issues/5

#6: Corrupt MqttSuback results in non-DecodeError exception.

    While decoding MqttSuback, a corrupted SubscribeResult results in
    a TypeError instead of DecodeError.  This violates the decode
    method's interface spec.

    https://github.com/kcallin/mqtt-codec/issues/6


1.0.1 (2018-11-28)
===================

New
----

#4: MqttConnect.__repr__ has seconds units on keep_alive.

    https://github.com/kcallin/haka-mqtt/issues/21


1.0.0 (2018-11-24)
===================

New
----

* First stable production release.


0.1.3 (2018-11-17)
===================

New
----

* Python 3 support.
* Updating packaging mechanism.


0.1.2 (2018-11-15)
===================

New
----
* Python 3 compatibility.
* MqttConnect object now has read-only attributes.

Fixes
------
#2: MqttPublish.payload somtimes has a type that is not bytes.

    https://github.com/kcallin/mqtt-codec/issues/2


0.1.1 (2018-10-22)
===================
* Documentation improvements.
* MqttConnect.__str__ no longer shows user/pass.
* setup.py:install_requires now compatible with setuptools 18.


0.1.0 (2018-10-03)
===================

Initial release.


tests/bench.py
from __future__ import print_function
from mqtt_codec import packet
from tests.stopwatch import Stopwatch


def bench_mqtt_connect_init(iterations=1000000):
    for i in xrange(0, iterations):
        packet.MqttConnect('client_id', False, 15000, username='username', password='password',
                           will=packet.MqttWill(0, 'topic', 'message', True))


def main():
    stopwatch = Stopwatch()

    with stopwatch:
        bench_mqtt_connect_init()

    print('bench_mqtt_connect_init', stopwatch.elapsed())


if __name__ == '__main__':
    main()


tests/test_io.py
import unittest
from io import BytesIO

from mqtt_codec.io import encode_bytes, decode_bytes, LimitReader, UnderflowDecodeError, encode_utf8, decode_utf8, \
    Utf8DecodeError, BytesReader


class TestDecodeBytes(unittest.TestCase):
    def test_len_underflow(self):
        with BytesIO() as buf:
            encode_bytes(b'01234', buf)
            buf.seek(0)
            limited_buf = LimitReader(buf, 1)
            self.assertRaises(UnderflowDecodeError, decode_bytes, limited_buf)

    def test_body_underflow(self):
        with BytesIO() as buf:
            encode_bytes(b'01234', buf)
            buf.seek(0)
            limited_buf = LimitReader(buf, 3)
            self.assertRaises(UnderflowDecodeError, decode_bytes, limited_buf)

    def test_encode_bytes_type_error(self):
        with BytesIO() as buf:
            self.assertRaises(TypeError, encode_bytes, 1, buf)


class TestDecodeUtf8(unittest.TestCase):
    def test_len_underflow(self):
        with BytesIO() as buf:
            encode_utf8('abcd', buf)
            buf.seek(0)
            limited_buf = LimitReader(buf, 1)
            self.assertRaises(UnderflowDecodeError, decode_utf8, limited_buf)

    def test_body_underflow(self):
        with BytesIO() as buf:
            encode_utf8('abcd', buf)
            buf.seek(0)
            limited_buf = LimitReader(buf, 3)
            self.assertRaises(UnderflowDecodeError, decode_utf8, limited_buf)

    def test_invalid_utf8(self):
        with BytesIO() as buf:
            encode_bytes(b'0\xff1234', buf)
            buf.seek(0)
            self.assertRaises(Utf8DecodeError, decode_utf8, buf)


class TestLimitReader(unittest.TestCase):
    def test_unlimited(self):
        with BytesIO() as buf:
            expected_bytes = b'12345'
            buf.write(expected_bytes)
            r = LimitReader(buf)
            buf.seek(0)
            bytes_read = r.read(None)
            self.assertEqual(expected_bytes, bytes_read)


class TestBytesReader(unittest.TestCase):
    def test_read_after_close(self):
        with BytesIO() as buf:
            r = BytesReader(b'12345')
            self.assertFalse(r.closed)
            r.close()
            self.assertTrue(r.closed)
            self.assertRaises(ValueError, r.read)

    def test_context(self):
        with BytesReader(b'asdfasdf') as r:
            self.assertFalse(r.closed)
        self.assertTrue(r.closed)

tests/stopwatch.py
from time import time


class Stopwatch(object):
    def __init__(self):
        self.__start_time = None
        self.__stop_time = None

    def __enter__(self):
        self.__start_time = time()
        return self

    def elapsed(self):
        if self.__start_time is not None:
            stop_time = self.__stop_time
            if stop_time is None:
                stop_time = time()
            rv = stop_time - self.__start_time
        else:
            rv = None

        return rv

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.__stop_time = time()
        return False


mqtt_codec/io.py
"""
========================
`mqtt_codec.io` Package
========================

A collection of helper functions and classes used to read MQTT control
packets.
"""

import codecs
from struct import Struct


def encode_bytes(src_buf, dst_file):
    """Encode a buffer length followed by the bytes of the buffer
    itself.

    Parameters
    ----------
    src_buf: bytes
        Source bytes to be encoded.  Function asserts that
        0 <= len(src_buf) <= 2**16-1.
    dst_file: file
        File-like object with write method.

    Returns
    -------
    int
        Number of bytes written to `dst_file`.
    """
    if not isinstance(src_buf, bytes):
        raise TypeError('src_buf must by bytes.')

    len_src_buf = len(src_buf)
    assert 0 <= len_src_buf <= 2**16-1
    num_written_bytes = len_src_buf + 2

    len_buf = FIELD_U16.pack(len_src_buf)
    dst_file.write(len_buf)
    dst_file.write(src_buf)

    return num_written_bytes


def decode_bytes(f):
    """Decode a buffer length from a 2-byte unsigned int then read the
    subsequent bytes.

    Parameters
    ----------
    f: file
        File-like object with read method.

    Raises
    ------
    UnderflowDecodeError
        When the end of stream is encountered before the end of the
        encoded bytes.

    Returns
    -------
    int
        Number of bytes read from `f`.
    bytes
        Value bytes decoded from `f`.
    """

    buf = f.read(FIELD_U16.size)
    if len(buf) < FIELD_U16.size:
        raise UnderflowDecodeError()

    (num_bytes,) = FIELD_U16.unpack_from(buf)
    num_bytes_consumed = FIELD_U16.size + num_bytes

    buf = f.read(num_bytes)
    if len(buf) < num_bytes:
        raise UnderflowDecodeError()

    return num_bytes_consumed, buf


def encode_utf8(s, f):
    """UTF-8 encodes string `s` to file-like object `f` according to
    the MQTT Version 3.1.1 specification in section 1.5.3.

    The maximum length for the encoded string is 2**16-1 (65535) bytes.
    An assertion error will result if the encoded string is longer.

    Parameters
    ----------
    s: str
        String to be encoded.
    f: file
        File-like object.

    Returns
    -------
    int
        Number of bytes written to f.
    """
    encode = codecs.getencoder('utf8')

    encoded_str_bytes, num_encoded_chars = encode(s)
    num_encoded_str_bytes = len(encoded_str_bytes)
    assert 0 <= num_encoded_str_bytes <= 2**16-1
    num_encoded_bytes = num_encoded_str_bytes + 2

    f.write(FIELD_U8.pack((num_encoded_str_bytes & 0xff00) >> 8))
    f.write(FIELD_U8.pack(num_encoded_str_bytes & 0x00ff))
    f.write(encoded_str_bytes)

    return num_encoded_bytes


def decode_utf8(f):
    """Decode a utf-8 string encoded as described in MQTT Version
    3.1.1 section 1.5.3 line 177.  This is a 16-bit unsigned length
    followed by a utf-8 encoded string.

    Parameters
    ----------
    f: file
        File-like object with read method.

    Raises
    ------
    UnderflowDecodeError
        Raised when a read failed to extract enough bytes from the
        underlying stream to decode the string.
    Utf8DecodeError
        When any code point in the utf-8 string is invalid.

    Returns
    -------
    int
        Number of bytes consumed.
    str
        A string utf-8 decoded from ``f``.
    """
    decode = codecs.getdecoder('utf8')

    buf = f.read(FIELD_U16.size)
    if len(buf) < FIELD_U16.size:
        raise UnderflowDecodeError()

    (num_utf8_bytes,) = FIELD_U16.unpack_from(buf)
    num_bytes_consumed = FIELD_U16.size + num_utf8_bytes

    buf = f.read(num_utf8_bytes)
    if len(buf) < num_utf8_bytes:
        raise UnderflowDecodeError()

    try:
        s, num_chars = decode(buf, 'strict')
    except UnicodeError as e:
        raise Utf8DecodeError(e)

    return num_bytes_consumed, s


def encode_varint(v, f):
    """Encode integer `v` to file `f`.

    Parameters
    ----------
    v: int
        Integer v >= 0.
    f: file
        Object containing a write method.

    Returns
    -------
    int
        Number of bytes written.
    """
    assert v >= 0
    num_bytes = 0

    while True:
        b = v % 0x80
        v = v // 0x80

        if v > 0:
            b = b | 0x80

        f.write(FIELD_U8.pack(b))

        num_bytes += 1
        if v == 0:
            break

    return num_bytes


def decode_varint(f, max_bytes=4):
    """Decode variable integer using algorithm similar to that described
    in MQTT Version 3.1.1 line 297.

    Parameters
    ----------
    f: file
        Object with a read method.
    max_bytes: int or None
        If a varint cannot be constructed using `max_bytes` or fewer
        from f then raises a `DecodeError`.  If None then there is no
        maximum number of bytes.

    Raises
    -------
    DecodeError
        When length is greater than max_bytes.
    UnderflowDecodeError
        When file ends before enough bytes can be read to construct the
        varint.

    Returns
    -------
    int
        Number of bytes consumed.
    int
        Value extracted from `f`.

    """
    num_bytes_consumed = 0

    value = 0
    m = 1

    while True:
        buf = f.read(1)
        if len(buf) == 0:
            raise UnderflowDecodeError()

        (u8,) = FIELD_U8.unpack(buf)
        value += (u8 & 0x7f) * m
        m *= 0x80
        num_bytes_consumed += 1

        if u8 & 0x80 == 0:
            # No further bytes
            break
        elif max_bytes is not None and num_bytes_consumed >= max_bytes:
            raise DecodeError('Variable integer contained more than maximum bytes ({}).'.format(max_bytes))

    return num_bytes_consumed, value


class DecodeError(Exception):
    pass


class UnderflowDecodeError(DecodeError):
    pass


class Utf8DecodeError(DecodeError):
    def __init__(self, e):
        assert isinstance(e, UnicodeError)
        self.error = e


class EncodeError(Exception):
    pass


class OverflowEncodeError(EncodeError):
    pass


class OversizePacketEncodeError(EncodeError):
    """Raised when the parameters used to create the MQTT packet would
    result in an impossibly large packet."""
    pass


class FileDecoder(object):
    """Creates an object that extracts values from the file-like
    object `f`.

    Parameters
    ----------
    f: file
        Object with read method.
    """

    def __init__(self, f):
        self.__f = f
        self.__num_bytes_consumed = 0

    @property
    def num_bytes_consumed(self):
        """int: number of bytes consumed from underlying stream."""
        return self.__num_bytes_consumed

    def unpack(self, struct):
        """Read as many bytes as are required to extract struct then
        unpack and return a tuple of the values.

        Raises
        ------
        UnderflowDecodeError
            Raised when a read failed to extract enough bytes from the
            underlying stream to extract the bytes.

        Parameters
        ----------
        struct: struct.Struct

        Returns
        -------
        tuple
            Tuple of extracted values.
        """
        v = struct.unpack(self.read(struct.size))
        return v

    def unpack_utf8(self):
        """Decode a utf-8 string encoded as described in MQTT Version
        3.1.1 section 1.5.3 line 177.  This is a 16-bit unsigned length
        followed by a utf-8 encoded string.

        Raises
        ------
        UnderflowDecodeError
            Raised when a read failed to extract enough bytes from the
            underlying stream to decode the string.
        DecodeError
            When any code point in the utf-8 string is invalid.

        Returns
        -------
        int
            Number of bytes consumed.
        str
            A string utf-8 decoded from the underlying stream.
        """
        num_bytes_consumed, s = decode_utf8(self.__f)
        self.__num_bytes_consumed += num_bytes_consumed
        return num_bytes_consumed, s

    def unpack_bytes(self):
        """Unpack a utf-8 string encoded as described in MQTT Version
        3.1.1 section 1.5.3 line 177.  This is a 16-bit unsigned length
        followed by a utf-8 encoded string.

        Returns
        -------
        int
            Number of bytes consumed
        bytes
            A bytes object extracted from the underlying stream.
        """
        num_bytes_consumed, b = decode_bytes(self.__f)
        self.__num_bytes_consumed += num_bytes_consumed
        return num_bytes_consumed, b

    def unpack_varint(self, max_bytes):
        """Decode variable integer using algorithm similar to that described
        in MQTT Version 3.1.1 line 297.

        Parameters
        ----------
        max_bytes: int or None
            If a varint cannot be constructed using `max_bytes` or fewer
            from f then raises a `DecodeError`.  If None then there is no
            maximum number of bytes.

        Raises
        -------
        DecodeError
            When length is greater than max_bytes.
        UnderflowDecodeError
            When file ends before enough bytes can be read to construct the
            varint.

        Returns
        -------
        int
            Number of bytes consumed.
        int
            Value extracted from `f`.

        """
        num_bytes_consumed, value = decode_varint(self.__f, max_bytes)
        self.__num_bytes_consumed += num_bytes_consumed
        return num_bytes_consumed, value

    def read(self, num_bytes):
        """Read `num_bytes` and return them.

        Parameters
        ----------
        num_bytes : int
            Number of bytes to extract from the underlying stream.

        Raises
        ------
        UnderflowDecodeError
            Raised when a read failed to extract enough bytes from the
            underlying stream to extract the bytes.

        Returns
        -------
        bytes
            A bytes object extracted from underlying stream.
        """
        buf = self.__f.read(num_bytes)
        assert len(buf) <= num_bytes
        if len(buf) < num_bytes:
            raise UnderflowDecodeError()
        self.__num_bytes_consumed += num_bytes

        return buf


class LimitReader(object):
    """Reads up to ``limit`` bytes from the underlying file.  If
    ``limit`` is none then reads to the end of the file.

    Parameters
    -----------
    f: file
        File-like object with read method.
    limit: int optional
        Maximum number of bytes to read from the underlying file or
        ``None`` the reader should continue until the end of the file.
    """

    def __init__(self, f, limit=None):
        self.__f = f
        self.__num_bytes_consumed = 0
        self.__limit = limit

    @property
    def limit(self):
        """int or None: maximum number of bytes to read from underlying stream."""
        return self.__limit

    def read(self, max_bytes=1):
        """Read at most `max_bytes` from internal buffer.

        Parameters
        -----------
        max_bytes: int
            Maximum number of bytes to read.

        Returns
        --------
        bytes
            Bytes extracted from internal buffer.  Length may be less
            than ``max_bytes``.  On end-of file returns a bytes object
            with zero-length.
        """

        if self.limit is None:
            b = self.__f.read(max_bytes)
        else:
            if self.__num_bytes_consumed + max_bytes > self.limit:
                max_bytes = self.limit - self.__num_bytes_consumed
            b = self.__f.read(max_bytes)
        self.__num_bytes_consumed += len(b)

        return b


class BytesReader(object):
    """Creates a file-like object that reads from a buffer.

    Parameters
    ----------
    buf: bytes or bytearray
        Object to read from.
    """

    def __init__(self, buf):
        assert isinstance(buf, (bytes, bytearray)), type(buf)
        self.__buf = buf
        self.__num_bytes_consumed = 0

    def read(self, max_bytes=1):
        """Read at most `max_bytes` from internal buffer.

        Parameters
        -----------
        max_bytes: int
            Maximum number of bytes to read.

        Raises
        ------
        ValueError
            If read is called after close has been called.

        Returns
        --------
        bytes
            Bytes extracted from internal buffer.  Length may be less
            than `max_bytes`.  On end-of file returns a bytes object
            with zero-length.
        """
        if self.__num_bytes_consumed is None:
            raise ValueError('I/O operation on closed file.')

        if self.__num_bytes_consumed + max_bytes >= len(self.__buf):
            max_bytes = len(self.__buf) - self.__num_bytes_consumed

        b = self.__buf[self.__num_bytes_consumed:self.__num_bytes_consumed + max_bytes]
        self.__num_bytes_consumed += max_bytes

        if isinstance(b, bytearray):
            b = bytes(b)

        assert isinstance(b, bytes)
        return b

    @property
    def closed(self):
        """bool: ``True`` if `self.close()` has been called; ``False`` otherwise."""
        return self.__num_bytes_consumed is None

    def close(self):
        """Read operations conducted after this method is called will
        raise `ValueError`.  This makes the object behave like other
        read objects even though no resources are freed."""
        self.__num_bytes_consumed = None

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()
        return False

FIELD_U16 = Struct('>H')
FIELD_PACKET_ID = FIELD_U16
FIELD_U8 = Struct('>B')

pyvertest.py
#!/bin/env python3
import sys
from os.path import (
    dirname,
    abspath,
    join, isdir)
from shutil import rmtree
from tempfile import mkdtemp

from plumbum import local

host_mqtt_codec_dir = abspath(dirname(__file__))
docker = local['docker']
git = local['git']


def run_cmd(image_id, host_mqtt_codec_dir, container_mqtt_codec_dir, cmd):
    container_id = docker('create', image_id, *cmd).strip()
    try:
        print('Created {container_id} from image {image_id}'.format(container_id=container_id, image_id=image_id))
        docker('cp', host_mqtt_codec_dir, '{}:{}'.format(container_id, container_mqtt_codec_dir))
        print('Cloned mqtt_codec to {container_id}:{container_mqtt_codec_dir}'.format(container_id=container_id,
                                                                          container_mqtt_codec_dir=container_mqtt_codec_dir))
        print('Running container', container_id)
        proc = docker.popen(args=['start', '-a', container_id], stdout=sys.stdout, stderr=sys.stderr)
        rc = proc.wait()
        print('Return code', rc)
    finally:
        print('Removing container id {}.'.format(container_id))
        docker('rm', container_id)

    return rc


def main():
    images = [
        'python:2.7-alpine3.8',
        'python:3.4',
        'python:3.5',
        'python:3.6-alpine3.8',
        'python:3.7-alpine3.8',
    ]

    return_codes = []
    container_mqtt_codec_dir= '/mqtt-codec'

    temp_host_mqtt_codec_dir = mkdtemp()
    assert isdir(temp_host_mqtt_codec_dir)
    try:
        git('clone', host_mqtt_codec_dir, temp_host_mqtt_codec_dir)
        print('Cloned {} into {}.'.format(host_mqtt_codec_dir, temp_host_mqtt_codec_dir))
        for image in images:
            print(50 * '*')
            return_codes.append(run_cmd(image, temp_host_mqtt_codec_dir, container_mqtt_codec_dir,
                                        ['python', join(container_mqtt_codec_dir, 'setup.py'), 'test']))
            return_codes.append(run_cmd(image, temp_host_mqtt_codec_dir, container_mqtt_codec_dir,
                                        ['pip', 'install', container_mqtt_codec_dir]))
    finally:
        print('Removing temp directory {}.'.format(temp_host_mqtt_codec_dir))
        rmtree(temp_host_mqtt_codec_dir)

    if any(return_codes):
        num_okay = len(rc for rc in return_codes if rc)
        num_fail = len(return_codes) - num_okay
        print('! {} okay, {} fail'.format(num_okay, num_fail))
    else:
        print('> {}/{} okay.'.format(len(return_codes), len(return_codes)))


if __name__ == '__main__':
    main()
